File: src/niepce/ui/moduleshell.cpp
/*
 * niepce - niepce/ui/moduleshell.cpp
 *
 * Copyright (C) 2007-2022 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <glibmm/i18n.h>
#include <glibmm/ustring.h>

#include <gtkmm/celllayout.h>
#include <gtkmm/cellrenderer.h>

#include "fwk/base/debug.hpp"
#include "engine/db/libfile.hpp"
#include "fwk/toolkit/application.hpp"
#include "fwk/toolkit/gtkutils.hpp"
#include "moduleshell.hpp"
#include "niepcewindow.hpp"
#include "metadatapanecontroller.hpp"

#include "rust_bindings.hpp"

namespace ui {

ModuleShell::~ModuleShell()
{
    m_widget = nullptr;
}

void ModuleShell::c_on_module_activated(GtkWidget*, const char* name, ModuleShell* self)
{
    self->on_module_activated(name);
}

void ModuleShell::c_on_module_deactivated(GtkWidget*, const char* name, ModuleShell* self)
{
    self->on_module_deactivated(name);
}

Gtk::Widget * ModuleShell::buildWidget()
{
    if(m_widget) {
        return m_widget;
    }

    m_widget = m_shell_widget;
    m_shell_widget->insert_action_group("shell", m_actionGroup);

    m_selection_controller = SelectionController_2::Ptr(new SelectionController_2(*m_libraryclient));
    add(m_selection_controller);

    m_menu = Gio::Menu::create();

    // "go-previous"
    fwk::add_menu_action(m_actionGroup.get(), "PrevImage",
                         sigc::mem_fun(*m_selection_controller->obj(),
                                       &SelectionController::select_previous),
                         m_menu, _("Back"), "shell", "Left");

    // "go-next"
    fwk::add_menu_action(m_actionGroup.get(), "NextImage",
                         sigc::mem_fun(*m_selection_controller->obj(),
                                  &SelectionController::select_next),
                         m_menu, _("Forward"), "shell", "Right");

    auto section = Gio::Menu::create();
    m_menu->append_section(section);

    // "object-rotate-left"
    fwk::add_menu_action(m_actionGroup.get(), "RotateLeft",
                         sigc::bind(
                             sigc::mem_fun(*m_selection_controller->obj(),
                                           &SelectionController::rotate), -90),
                         section, _("Rotate Left"), "shell", "bracketleft");

    // "object-rotate-right"
    fwk::add_menu_action(m_actionGroup.get(), "RotateRight",
                         sigc::bind(
                             sigc::mem_fun(*m_selection_controller->obj(),
                                           &SelectionController::rotate), 90),
                         section, _("Rotate Right"), "shell", "bracketright");

    section = Gio::Menu::create();
    m_menu->append_section(section);

    auto submenu = Gio::Menu::create();
    section->append_submenu(_("Set Label"), submenu);

    fwk::add_menu_action(m_actionGroup.get(),
                         "SetLabel6",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_label),
                                    1),
                         submenu, _("Label 6"), "shell", "<Primary>6");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetLabel7",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_label),
                                    2),
                         submenu, _("Label 7"), "shell", "<Primary>7");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetLabel8",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_label),
                                    3),
                         submenu, _("Label 8"), "shell", "<Primary>8");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetLabel9",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_label),
                                    4),
                         submenu, _("Label 9"), "shell", "<Primary>9");

    submenu = Gio::Menu::create();
    section->append_submenu(_("Set Rating"), submenu);

    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating0",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_rating),
                                    0),
                         submenu, _("Unrated"), "shell", "<Primary>0");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating1",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                             &SelectionController::set_rating),
                                    1),
                         submenu, _("Rating 1"), "shell", "<Primary>1");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating2",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_rating),
                                    2),
                         submenu, _("Rating 2"), "shell", "<Primary>2");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating3",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_rating),
                                    3),
                         submenu, _("Rating 3"), "shell", "<Primary>3");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating4",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_rating),
                                    4),
                         submenu, _("Rating 4"), "shell", "<Primary>4");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetRating5",
                         sigc::bind(sigc::mem_fun(*m_selection_controller->obj(),
                                                  &SelectionController::set_rating),
                                    5),
                         submenu, _("Rating 5"), "shell", "<Primary>5");

    submenu = Gio::Menu::create();
    section->append_submenu(_("Set Flag"), submenu);

    fwk::add_menu_action(m_actionGroup.get(),
                         "SetFlagReject",
                         sigc::bind(
                             sigc::mem_fun(*m_selection_controller->obj(),
                                           &SelectionController::set_flag),
                             -1),
                         submenu, _("Flag as Rejected"), "shell", "<Primary><Shift>x");
    fwk::add_menu_action(m_actionGroup.get(),
                         "SetFlagNone",
                         sigc::bind(
                             sigc::mem_fun(*m_selection_controller->obj(),
                                           &SelectionController::set_flag),
                             0),
                         submenu, _("Unflagged"), "shell", "<Primary><Shift>u");
    fwk::add_menu_action(m_actionGroup.get(),
                          "SetFlagPick",
                         sigc::bind(
                             sigc::mem_fun(*m_selection_controller->obj(),
                                           &SelectionController::set_flag),
                             1),
                         submenu, _("Flag as Pick"), "shell", "<Primary><Shift>p");

    section = Gio::Menu::create();
    m_menu->append_section(section);

    fwk::add_menu_action(m_actionGroup.get(),
                         "WriteMetadata",
                         sigc::mem_fun(*m_selection_controller->obj(),
                                       &SelectionController::write_metadata),
                         section, _("Write metadata"), "shell", nullptr);

    // Module menu placeholder
    m_module_menu = Gio::Menu::create();
    m_menu->append_section(m_module_menu);

    gtk_menu_button_set_menu_model(
        GTK_MENU_BUTTON(m_shell->getMenuButton()), G_MENU_MODEL(m_menu->gobj()));

    m_gridview = GridViewModule::Ptr(
        new GridViewModule(*this, m_selection_controller->obj()->get_list_store().clone()));
    add_library_module(m_gridview, "grid", _("Catalog"));

    m_selection_controller->add_selectable(m_gridview);
    m_selection_controller->obj()->add_selected_listener(
        std::make_unique<npc::SelectionListener>([this] (eng::library_id_t id) {
            this->on_image_selected(id);
        }));
    m_selection_controller->obj()->add_activated_listener(
        std::make_unique<npc::SelectionListener>([this] (eng::library_id_t id) {
            this->on_image_activated(id);
        }));

    m_darkroom = dr::DarkroomModule::Ptr(new dr::DarkroomModule(*this));
    add_library_module(m_darkroom, "darkroom", _("Darkroom"));

    m_mapm = mapm::MapModule::Ptr(new mapm::MapModule(*this));
    add_library_module(m_mapm, "map", _("Map"));

    g_signal_connect(G_OBJECT(m_shell_widget->gobj()), "activated", G_CALLBACK(ModuleShell::c_on_module_activated), this);
    g_signal_connect(G_OBJECT(m_shell_widget->gobj()), "deactivated", G_CALLBACK(ModuleShell::c_on_module_deactivated), this);

    // TODO PrintModuleController
    // add_library_module(, _("Print"));
    return m_widget;
}

void ModuleShell::action_edit_delete()
{
    DBG_OUT("shell - delete");
    m_selection_controller->obj()->move_to_trash();
}

void ModuleShell::add_library_module(const ILibraryModule::Ptr & module,
                                     const std::string & name,
                                     const std::string & label)
{
    auto w = module->buildWidget();
    if(w) {
        add(module);
        m_shell->appendPage((char*)w->gobj(), name, label);
        m_modules.insert(std::make_pair(name, module));
    }
}

void ModuleShell::on_ready()
{
}

void ModuleShell::on_content_will_change()
{
    m_selection_controller->obj()->content_will_change();
}

void ModuleShell::on_image_selected(eng::library_id_t id)
{
    DBG_OUT("selected callback %Ld", (long long)id);
    if(id > 0) {
        ffi::libraryclient_request_metadata(&m_libraryclient->client(), id);
    }
    else  {
        m_gridview->display_none();
    }
}

void ModuleShell::on_image_activated(eng::library_id_t id)
{
    DBG_OUT("on image activated %Ld", (long long)id);
    auto& store = m_selection_controller->obj()->get_list_store();
    auto libfile = ImageListStore_get_file(store.unwrap_ref(), id);
    if (libfile) {
        m_darkroom->set_image(std::optional(std::move(libfile)));
        m_shell->activatePage("darkroom");
    }
}

void ModuleShell::on_module_deactivated(const std::string & name) const
{
    auto module = m_modules.find(name);
    if (module != m_modules.end()) {
        m_module_menu->remove_all();
        module->second->set_active(false);
    }
}

void ModuleShell::on_module_activated(const std::string & name) const
{
    auto module = m_modules.find(name);
    if (module != m_modules.end()) {
        auto menu = module->second->getMenu();
        if (menu) {
            m_module_menu->append_section(menu);
        }
        module->second->set_active(true);
    }
}

void
ModuleShell::on_lib_notification(const eng::LibNotification &ln)
{
    m_gridview->on_lib_notification(ln);
    m_mapm->on_lib_notification(ln);
    m_selection_controller->obj()->on_lib_notification(ln, m_libraryclient->thumbnailCache());
}


}

/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:80
  End:
*/

File: niepce-main/src/niepce/ui/module_shell.rs
/*
 * niepce - niepce/ui/module_shell.rs
 *
 * Copyright (C) 2022-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
use std::sync::Weak;

use gettextrs::gettext as i18n;
use gtk4::prelude::*;
use npc_fwk::{gio, glib, gtk4};

use super::{
    GridViewModule, ImageListStore, LibraryModule, ModuleShellWidget, SelectionController,
};
use crate::NiepceApplication;
use crate::modules::{DarkroomModule, MapModule};
use npc_engine::catalog;
use npc_engine::library::notification::LibNotification;
use npc_engine::libraryclient::{ClientInterface, LibraryClientHost};
use npc_fwk::dbg_out;
use npc_fwk::send_async_local;
use npc_fwk::toolkit::gtk_utils::add_menu_action;
use npc_fwk::toolkit::{Controller, ControllerImplCell, Sender, UiController};

pub enum Event {
    ModuleActivated(String),
    ModuleDeactivated(String),
}

pub struct ModuleShell {
    imp_: ControllerImplCell<Event, ()>,
    widget: ModuleShellWidget,
    action_group: gio::SimpleActionGroup,
    selection_controller: Rc<SelectionController>,
    // currently a proxy that will bridge the C++ implementation
    gridview: Rc<GridViewModule>,
    mapm: Rc<MapModule>,
    darkroom: Rc<DarkroomModule>,
    menu: gio::Menu,
    module_menu: gio::Menu,
    client: Rc<LibraryClientHost>,
    modules: RefCell<HashMap<String, Rc<dyn LibraryModule>>>,
}

impl ModuleShell {
    pub fn new(
        client_host: &Rc<LibraryClientHost>,
        app: Weak<NiepceApplication>,
    ) -> Rc<ModuleShell> {
        let selection_controller = SelectionController::new(client_host, app);
        let menu = gio::Menu::new();
        let shell = Rc::new(ModuleShell {
            imp_: ControllerImplCell::default(),
            widget: ModuleShellWidget::new(),
            action_group: gio::SimpleActionGroup::new(),
            gridview: GridViewModule::new(&selection_controller, &menu, client_host),
            mapm: MapModule::new(),
            darkroom: DarkroomModule::new(client_host),
            selection_controller,
            menu,
            module_menu: gio::Menu::new(),
            client: client_host.clone(),
            modules: RefCell::new(HashMap::default()),
        });

        <Self as Controller>::start(&shell);

        shell
            .widget
            .insert_action_group("shell", Some(&shell.action_group));

        Self::build_gridview_context_menu(&shell);
        shell.widget.menu_button().set_menu_model(Some(&shell.menu));

        shell.add_library_module(&shell.gridview, "grid", &i18n("Catalog"));
        let sender = shell.selection_controller.sender();
        shell
            .gridview
            .image_grid_view
            .connect_activate(glib::clone!(
                #[strong]
                sender,
                move |_, pos| {
                    send_async_local!(
                        super::selection_controller::SelectionInMsg::Activated(pos),
                        sender
                    )
                }
            ));

        shell
            .selection_controller
            .set_forwarder(Some(Box::new(glib::clone!(
                #[weak]
                shell,
                move |msg| {
                    use super::selection_controller::SelectionOutMsg;
                    match msg {
                        SelectionOutMsg::Selected(id) => shell.on_image_selected(id),
                        SelectionOutMsg::Activated(id) => shell.on_image_activated(id),
                    }
                }
            ))));

        // built-in modules;
        shell.add_library_module(&shell.darkroom, "darkroom", &i18n("Darkroom"));
        shell.add_library_module(&shell.mapm, "map", &i18n("Map"));

        let tx = shell.sender();
        shell.widget.connect(
            "activated",
            true,
            glib::clone!(
                #[strong]
                tx,
                move |value| {
                    let name = value[1]
                        .get::<&str>()
                        .expect("Failed to convert callback parameter")
                        .to_string();
                    npc_fwk::send_async_local!(Event::ModuleActivated(name), tx);
                    None
                }
            ),
        );
        let tx = shell.sender();
        shell.widget.connect(
            "deactivated",
            true,
            glib::clone!(
                #[strong]
                tx,
                move |value| {
                    let name = value[1]
                        .get::<&str>()
                        .expect("Failed to convert callback parameter")
                        .to_string();
                    npc_fwk::send_async_local!(Event::ModuleDeactivated(name), tx);
                    None
                }
            ),
        );
        shell
    }

    /// Build the GridView context menu.
    fn build_gridview_context_menu(shell: &Rc<Self>) {
        let group = shell.action_group.upcast_ref::<gio::ActionMap>();
        add_menu_action(
            group,
            "PrevImage",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.select_previous()
            ),
            &shell.menu,
            Some(&i18n("Back")),
            Some("shell"),
            Some("Left"),
        );
        add_menu_action(
            group,
            "NextImage",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.select_next()
            ),
            &shell.menu,
            Some(&i18n("Forward")),
            Some("shell"),
            Some("Right"),
        );

        let section = gio::Menu::new();
        shell.menu.append_section(None, &section);
        add_menu_action(
            group,
            "RotateLeft",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.rotate(-90)
            ),
            &section,
            Some(&i18n("Rotate Left")),
            Some("shell"),
            Some("bracketleft"),
        );
        add_menu_action(
            group,
            "RotateRight",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.rotate(90)
            ),
            &section,
            Some(&i18n("Rotate Right")),
            Some("shell"),
            Some("bracketright"),
        );

        let section = gio::Menu::new();
        shell.menu.append_section(None, &section);
        let submenu = gio::Menu::new();
        section.append_submenu(Some(&i18n("Set Label")), &submenu);

        add_menu_action(
            group,
            "SetLabel6",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_label(1)
            ),
            &submenu,
            Some(&i18n("Label 6")),
            Some("shell"),
            Some("<Primary>6"),
        );
        add_menu_action(
            group,
            "SetLabel7",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_label(2)
            ),
            &submenu,
            Some(&i18n("Label 7")),
            Some("shell"),
            Some("<Primary>7"),
        );
        add_menu_action(
            group,
            "SetLabel8",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_label(3)
            ),
            &submenu,
            Some(&i18n("Label 8")),
            Some("shell"),
            Some("<Primary>8"),
        );
        add_menu_action(
            group,
            "SetLabel9",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_label(4)
            ),
            &submenu,
            Some(&i18n("Label 9")),
            Some("shell"),
            Some("<Primary>9"),
        );

        let submenu = gio::Menu::new();
        section.append_submenu(Some(&i18n("Set Rating")), &submenu);
        add_menu_action(
            group,
            "SetRating0",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(0)
            ),
            &submenu,
            Some(&i18n("Unrated")),
            Some("shell"),
            Some("<Primary>0"),
        );
        add_menu_action(
            group,
            "SetRating1",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(1)
            ),
            &submenu,
            Some(&i18n("Rating 1")),
            Some("shell"),
            Some("<Primary>1"),
        );
        add_menu_action(
            group,
            "SetRating2",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(2)
            ),
            &submenu,
            Some(&i18n("Rating 2")),
            Some("shell"),
            Some("<Primary>2"),
        );
        add_menu_action(
            group,
            "SetRating3",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(3)
            ),
            &submenu,
            Some(&i18n("Rating 3")),
            Some("shell"),
            Some("<Primary>3"),
        );
        add_menu_action(
            group,
            "SetRating4",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(4)
            ),
            &submenu,
            Some(&i18n("Rating 4")),
            Some("shell"),
            Some("<Primary>4"),
        );
        add_menu_action(
            group,
            "SetRating5",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_rating(5)
            ),
            &submenu,
            Some(&i18n("Rating 5")),
            Some("shell"),
            Some("<Primary>5"),
        );

        let submenu = gio::Menu::new();
        section.append_submenu(Some(&i18n("Set Flag")), &submenu);
        add_menu_action(
            group,
            "SetFlagReject",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_flag(-1)
            ),
            &submenu,
            Some(&i18n("Flag as Rejected")),
            Some("shell"),
            Some("<Primary><Shift>x"),
        );
        add_menu_action(
            group,
            "SetFlagNone",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_flag(0)
            ),
            &submenu,
            Some(&i18n("Unflagged")),
            Some("shell"),
            Some("<Primary><Shift>u"),
        );
        add_menu_action(
            group,
            "SetFlagPick",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.set_flag(1)
            ),
            &submenu,
            Some(&i18n("Flag as Pick")),
            Some("shell"),
            Some("<Primary><Shift>p"),
        );

        let section = gio::Menu::new();
        shell.menu.append_section(None, &section);
        add_menu_action(
            group,
            "WriteMetadata",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.write_metadata()
            ),
            &section,
            Some(&i18n("Write metadata")),
            Some("shell"),
            None,
        );

        let section = gio::Menu::new();
        shell.menu.append_section(None, &section);
        add_menu_action(
            group,
            "Delete",
            glib::clone!(
                #[weak(rename_to = selection_controller)]
                shell.selection_controller,
                move |_, _| selection_controller.delete_from_view()
            ),
            &section,
            Some(&i18n("Delete")),
            Some("shell"),
            None,
        );

        shell.menu.append_section(None, &shell.module_menu);
    }

    pub fn selection_sender(&self) -> Sender<<SelectionController as Controller>::InMsg> {
        self.selection_controller.sender()
    }

    pub fn image_list_store(&self) -> Rc<ImageListStore> {
        self.selection_controller.list_store().clone()
    }

    pub fn on_lib_notification(&self, ln: &LibNotification) {
        self.gridview.on_lib_notification(ln, self.client.client());
        self.darkroom.on_lib_notification(ln);
        self.mapm.on_lib_notification(ln);
        self.selection_controller
            .on_lib_notification(ln, self.client.thumbnail_cache());
    }

    pub fn action_edit_delete(&self) {
        self.selection_controller.move_to_trash();
    }

    fn add_library_module<T: LibraryModule + 'static>(
        &self,
        module: &Rc<T>,
        name: &str,
        label: &str,
    ) {
        let widget = module.widget();
        self.widget.append_page(widget, name, label);
        self.modules
            .borrow_mut()
            .insert(name.to_string(), module.clone());
    }

    pub fn on_content_will_change(&self, content: super::ContentView) {
        self.selection_controller.content_will_change(content);
    }

    fn on_image_selected(&self, id: catalog::LibraryId) {
        dbg_out!("Selected callback for {}", id);
        if id > 0 {
            self.client.client().request_metadata(id);
        } else {
            self.gridview.display_none()
        }
        // Forward to the darkroom module.
        let store = &self.selection_controller.list_store();
        self.darkroom.set_image(store.file(id).as_ref());
    }

    fn on_image_activated(&self, id: catalog::LibraryId) {
        dbg_out!("Activated callback for {}", id);
        let store = &self.selection_controller.list_store();
        self.darkroom.set_image(store.file(id).as_ref());
        self.widget.activate_page("darkroom");
    }

    fn module_activated(&self, name: &str) {
        if let Some(module) = self.modules.borrow().get(name) {
            if let Some(menu) = module.menu() {
                self.module_menu.append_section(None, menu);
            }
            module.set_active(true);
        }
    }

    fn module_deactivated(&self, name: &str) {
        if let Some(module) = self.modules.borrow().get(name) {
            self.module_menu.remove_all();
            module.set_active(false);
        }
    }
}

impl Controller for ModuleShell {
    type InMsg = Event;
    type OutMsg = ();

    npc_fwk::controller_imp_imp!(imp_);

    fn dispatch(&self, e: Event) {
        use Event::*;
        match e {
            ModuleActivated(ref name) => self.module_activated(name),
            ModuleDeactivated(ref name) => self.module_deactivated(name),
        }
    }
}

impl UiController for ModuleShell {
    fn widget(&self) -> &gtk4::Widget {
        self.widget.upcast_ref()
    }

    fn actions(&self) -> Option<(&str, &gio::ActionGroup)> {
        Some(("shell", self.action_group.upcast_ref()))
    }
}
